usage/usage.go

```diff
@@ -0,0 +1,132 @@
+	res := t.Publish(ctx, &pubsub.Message{Data: msg})
+	id, err := res.Get(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to publish message: %w", err)
+	}
+	log.V(2).Infof("Published event %s with message ID: %s", event.GetUuid(), id)
+	return nil
+}
+
+// ReportDeployClusterStart reports that a cluster deployment has started. The UUID of the
+// reported event is returned along with an error if the event reporting failed.
+func (r *Reporter) ReportDeployClusterStart(ctx context.Context, c *epb.Cluster) (string, error) {
+	event := newEvent()
+	event.Event = &epb.KNEEvent_DeployClusterStart{DeployClusterStart: &epb.DeployClusterStart{Cluster: c}}
+	if err := r.publishEvent(ctx, event); err != nil {
+		return "", fmt.Errorf("failed to report DeployClusterStart event with ID %s: %w", event.GetUuid(), err)
+	}
+	log.V(1).Infof("Reported DeployClusterStart event with ID %s", event.GetUuid())
+	return event.GetUuid(), nil
+}
+
+// ReportDeployClusterEnd reports that a cluster deployment has ended.
+func (r *Reporter) ReportDeployClusterEnd(ctx context.Context, id string, deployErr error) error {
+	event := newEvent()
+	event.Uuid = id
+	end := &epb.DeployClusterEnd{}
+	if deployErr != nil {
+		end.Error = deployErr.Error()
+	}
+	event.Event = &epb.KNEEvent_DeployClusterEnd{DeployClusterEnd: end}
+	if err := r.publishEvent(ctx, event); err != nil {
+		return fmt.Errorf("failed to report DeployClusterEnd event with ID %s: %w", event.GetUuid(), err)
+	}
+	log.V(1).Infof("Reported DeployClusterEnd event with ID %s", event.GetUuid())
+	return nil
+}
+
+// ReportCreateTopologyStart reports that a topology creation has started. The UUID of the
+// reported event is returned along with an error if the event reporting failed.
+func (r *Reporter) ReportCreateTopologyStart(ctx context.Context, t *epb.Topology) (string, error) {
+	event := newEvent()
+	event.Event = &epb.KNEEvent_CreateTopologyStart{CreateTopologyStart: &epb.CreateTopologyStart{Topology: t}}
+	if err := r.publishEvent(ctx, event); err != nil {
+		return "", fmt.Errorf("failed to report CreateTopologyStart event with ID %s: %w", event.GetUuid(), err)
+	}
+	log.V(1).Infof("Reported CreateTopologyStart event with ID %s", event.GetUuid())
+	return event.GetUuid(), nil
+}
+
+// ReportCreateTopologyEnd reports that a topology creation has ended.
+func (r *Reporter) ReportCreateTopologyEnd(ctx context.Context, id string, createErr error) error {
+	event := newEvent()
+	event.Uuid = id
+	end := &epb.CreateTopologyEnd{}
+	if createErr != nil {
+		end.Error = createErr.Error()
+	}
+	event.Event = &epb.KNEEvent_CreateTopologyEnd{CreateTopologyEnd: end}
+	if err := r.publishEvent(ctx, event); err != nil {
+		return fmt.Errorf("failed to report CreateTopologyEnd event with ID %s: %w", event.GetUuid(), err)
+	}
+	log.V(1).Infof("Reported CreateTopologyEnd event with ID %s", event.GetUuid())
+	return nil
+}
+
+func newEvent() *epb.KNEEvent {
+	return &epb.KNEEvent{
+		Uuid:      uuid.New(),
```

Comment:
Perhaps the id should be passed in rather than creating a UUID here.  ReportDeployClusterStart can then call `newEvent(uuid.New())` and make it clear it is creating a new UUID rather than using an existing one.

Reply:
done